rasa_core/policies/two_stage_fallback.py

```diff
@@ -0,0 +1,171 @@
+        continues as if the user had this intent from the beginning.
+        If the clarification was not recognised with high confidence, the user
+        is asked to confirm the recognized intent.
+        If the user confirms the intent, the story continues as if the user had
+        this intent from the beginning.
+        If the user denies, an ultimate fallback action is triggered
+        (e.g. a handoff to a human).
+    """
+
+    def __init__(self,
+                 nlu_threshold: float = 0.3,
+                 core_threshold: float = 0.3,
+                 fallback_action_name: Text = ACTION_DEFAULT_FALLBACK_NAME,
+                 ) -> None:
+        """Create a new Two Stage Fallback policy.
+
+                Args:
+                    nlu_threshold: minimum threshold for NLU confidence.
+                        If intent prediction confidence is lower than this,
+                        predict fallback action with confidence 1.0.
+                    core_threshold: if NLU confidence threshold is met,
+                        predict fallback action with confidence
+                        `core_threshold`. If this is the highest confidence in
+                        the ensemble, the fallback action will be executed.
+                    fallback_action_name: This action is executed if the user
+                        denies the recognised intent for the second time.
+                """
+        super(TwoStageFallbackPolicy, self).__init__()
+
+        self.nlu_threshold = nlu_threshold
+        self.core_threshold = core_threshold
+        self.fallback_action_name = fallback_action_name
+
+    def predict_action_probabilities(self,
+                                     tracker: DialogueStateTracker,
+                                     domain: Domain) -> List[float]:
+        """Predicts the next action if NLU confidence is low.
+        """
+
+        if (USER_INTENT_CONFIRM not in domain.intents or
+                USER_INTENT_DENY not in domain.intents):
+            raise InvalidDomain('The intents {} and {} must be present in the '
+                                'domain file to use this policy.'.format(
+                                    USER_INTENT_CONFIRM, USER_INTENT_CONFIRM))
+
+        nlu_data = tracker.latest_message.parse_data
+        nlu_confidence = nlu_data["intent"].get("confidence", 1.0)
+        last_intent_name = nlu_data['intent'].get('name', None)
+        should_fallback = self.should_fallback(nlu_confidence,
+                                               tracker.latest_action_name)
+        user_clarified = has_user_clarified(tracker)
+
+        if self.__is_user_input_expected(tracker):
+            result = confidence_scores_for('action_listen', FALLBACK_SCORE,
+                                           domain)
+        elif _has_user_denied(last_intent_name, tracker):
+            logger.debug("User denied suggested intent.")
```

Comment:
let's try to make these logging messages more informative by including e.g. the sender id and the intent. Imagine a log where 10 users chat with the bot, these messages won't help to debug in that case.